File: solutions/c/acronym/2/acronym.c
#include "acronym.h"

#include <string.h>
#include <malloc.h>

typedef enum state_t {
    SPACE,
    CAPITAL,
    CONSUME
} state_t;

#define INSERT(buff,k,c) buff[k]=c; k++;

char *abbreviate(const char *phrase){
    if (!phrase)
        return NULL;
    const size_t len = strlen(phrase);
    size_t count = 0;
    if (!len)
        return NULL;
    char *abbrev = (char*)calloc(len, sizeof(char));
    state_t state = SPACE;
    for(size_t i=0; i<len; i++){
        char ch = phrase[i];
        if (ch == ' ' || ch == '_' || ch == '-'){
            state = SPACE;
        } 
        else if(ch>='A' && ch <= 'Z'){
            if(state == CONSUME || state == SPACE){
                INSERT(abbrev, count, ch)}
            state = CAPITAL;    
        }
        else if(ch>='a' && ch <= 'z'){
            if(state == SPACE) {
                INSERT(abbrev, count, ch-'a'+'A')
                }
            state = CONSUME;
        }
     }
    return abbrev;
}

File: solutions/free-pascal/acronym/1/Acronym.pas
unit Acronym;

{$mode ObjFPC}{$H+}

interface

function abbreviate(const phrase: string) : string;

implementation

uses SysUtils;

type TState = (SPACE, CAPITAL, CONSUME);

function abbreviate(const phrase: string) : string;
var 
  state : TState;
  c : char;
begin
  result := '';
  for c in phrase do
  begin
    if (c=' ') or (c='_') or (c='-') then
      state := SPACE
    else if (c>='A') and (c<='Z') then
    begin
      if (state=SPACE) or (state=CONSUME) then
        result := result + c;
      state := CAPITAL;  
    end
    else if (c>='a') and (c<='z') then
    begin
      if (state=SPACE) then
        result := result + Chr(ord(c) - ord('a') + ord('A'));
      state := CONSUME;
    end
    else
      continue;
  end;

end;

end.
